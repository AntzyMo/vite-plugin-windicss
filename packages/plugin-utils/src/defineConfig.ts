import { WindiCssOptions } from './options'

export function defineConfig(config: WindiCssOptions) {
  return config
}
