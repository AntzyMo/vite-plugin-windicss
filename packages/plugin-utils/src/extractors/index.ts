export * from './default'
export * from './helper'
export * from './pug'
export * from './svelte'
