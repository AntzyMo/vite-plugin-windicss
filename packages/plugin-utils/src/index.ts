export * from './constants'
export * from './defineConfig'
export * from './createUtils'
export * from './options'
export * from './transforms'
export * from './extractors'
export * from './utils'
export * from './resolveOptions'
export * from './merge'
