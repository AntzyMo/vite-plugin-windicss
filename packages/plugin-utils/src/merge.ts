import type { FullConfig } from 'windicss/types/interfaces'

const isObject = (val: any) => toString.call(val) === '[object Object]'

function deepMerge(a: any, b: any, rootPath: string) {
  a = { ...a }
  Object.keys(b).forEach((key) => {
    if (isObject(a[key]))
      a[key] = deepMerge(a[key], b[key], rootPath ? `${rootPath}.${key}` : key)
    else if (Array.isArray(a[key]))
      a[key] = [...a[key], ...b[key]]
    else
      a[key] = b[key]
  })
  return a
}

export function mergeWindicssConfig(a: FullConfig, b: FullConfig) {
  // TODO: handle more special props
  return deepMerge(a, b, '')
}
