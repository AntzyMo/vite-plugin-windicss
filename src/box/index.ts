import { promises as fs, existsSync } from 'fs'
import { join, resolve } from 'path'
import WindiCssProcessor from 'windicss'
import { StyleSheet } from 'windicss/utils/style'
import { CSSParser } from 'windicss/utils/parser'
import fg from 'fast-glob'
import _debug from 'debug'
import { regexQuotedString, regexClassSplitter, regexClassCheck, regexHtmlTag, preflightTags, htmlTags, defaultAlias, TagNames, regexClassGroup } from './constants'
import { resolveOptions, WindiCssOptions, WindiBoxOptions } from './options'

import { toArray, kebabCase, include, exclude } from './utils'

export type WindiBox = ReturnType<typeof createBox>

export { preflightTags, htmlTags, defaultAlias }
export type { WindiBoxOptions, TagNames }

export function createBox(_options: WindiBoxOptions = {}) {
  const options = resolveOptions(_options)

  const {
    name,
    config,
    scan: enabledScan,
    scanOptions,
    transformCSS: enableCssTransform,
    preflight: enablePreflight,
    preflightOptions,
    sortUtilities,
    root,
    safelist,
  } = options

  const debug = {
    config: _debug(`${name}:config`),
    debug: _debug(`${name}:debug`),
    compile: _debug(`${name}:compile`),
    glob: _debug(`${name}:glob`),
    detect: _debug(`${name}:detect`),
  }

  let processor: WindiCssProcessor
  let configFilePath: string | undefined

  const regexId = new RegExp(`\\.(?:${scanOptions.fileExtensions.join('|')})$`, 'i')

  const configSafelist = new Set<string>()

  const classesGenerated = new Set<string>()
  const classesPending = new Set<string>()
  const tagsGenerated = new Set<string>()
  const tagsPending = new Set<string>()
  const tagsAvailable = new Set<string>()

  function loadConfiguration() {
    let resolved: WindiCssOptions = {}
    if (typeof config === 'string') {
      const path = resolve(root, config)
      if (!existsSync(path)) {
        console.warn(`[${name}] config file "${config}" not found, ignored`)
      }
      else {
        try {
          delete require.cache[require.resolve(path)]
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          resolved = require(path)
          configFilePath = path
          configSafelist.clear()
          // @ts-expect-error
          include(configSafelist, resolved?.purge?.options?.safelist || resolved?.purge?.options?.whitelist || [])
        }
        catch (e) {
          console.error(`[${name}] failed to load config "${config}"`)
          console.error(`[${name}] ${e.toString()}`)
          process.exit(1)
        }
      }
    }
    else {
      resolved = config
    }

    debug.config(JSON.stringify(resolved, null, 2))
    return resolved
  }

  function initWindicss() {
    return new WindiCssProcessor(loadConfiguration())
  }

  let _searching: Promise<void> | null

  async function scan() {
    if (!_searching) {
      const { fileExtensions, dirs, include, exclude } = scanOptions
      _searching = (async() => {
        const globs = dirs.map(i => join(i, `**/*.{${fileExtensions.join(',')}}`).replace(/\\/g, '/'))
        globs.unshift('index.html')
        globs.unshift(...include)

        debug.glob('globs', globs)

        const files = await fg(
          globs,
          {
            cwd: root,
            ignore: ['node_modules', '.git', ...exclude],
            onlyFiles: true,
            absolute: true,
          },
        )

        files.sort()

        debug.glob('files', files)

        const contents = await Promise.all(
          files
            .filter(id => isDetectTarget(id))
            .map(id => fs.readFile(id, 'utf-8')),
        )

        for (const content of contents)
          extractFile(content)
      })()
    }

    return _searching
  }

  function isDetectTarget(id: string) {
    return id.match(regexId)
  }

  function extractFile(code: string) {
    // classes
    Array.from(code.matchAll(regexQuotedString))
      .flatMap(m => (m[2] || '')
        .replace(regexClassGroup, (v) => {
          if (!classesGenerated.has(v))
            classesPending.add(v)
          return ''
        })
        .split(regexClassSplitter),
      )
      .filter(i => i.match(regexClassCheck))
      .forEach((i) => {
        if (!i || classesGenerated.has(i))
          return
        classesPending.add(i)
      })

    if (enablePreflight) {
      // preflight
      Array.from(code.matchAll(regexHtmlTag))
        .flatMap(([, i]) => i)
        .forEach((i) => {
          if (!tagsAvailable.has(i))
            i = preflightOptions.alias[kebabCase(i)]
          if (!tagsAvailable.has(i))
            return
          tagsPending.add(i)
          tagsAvailable.delete(i)
        })
    }

    debug.detect('classes', classesPending)
    debug.detect('tags', tagsPending)
  }

  function transformCSS(css: string) {
    if (!enableCssTransform)
      return css
    const style = new CSSParser(css, processor).parse()
    return style.build()
  }

  let style: StyleSheet = new StyleSheet()
  let _cssCache: string | undefined

  async function generateCSS() {
    if (enabledScan && scanOptions.runOnStartup)
      await scan()

    let changed = false

    if (classesPending.size) {
      const result = processor.interpret(Array.from(classesPending).join(' '))
      if (result.success.length) {
        include(classesGenerated, result.success)
        classesPending.clear()
        debug.compile(`compiled ${result.success.length} classes`)
        debug.compile(result.success)

        style = style.extend(result.styleSheet)
        changed = true
      }
    }

    if (enablePreflight && tagsPending.size) {
      const preflightStyle = processor.preflight(
        Array.from(tagsPending).map(i => `<${i}`).join(' '),
        preflightOptions.includeBase,
        preflightOptions.includeGlobal,
        preflightOptions.includePlugin,
      )
      style = style.extend(preflightStyle, true)
      include(tagsGenerated, tagsPending)
      tagsPending.clear()
      changed = true
    }

    if (changed || !_cssCache) {
      if (sortUtilities)
        style.sort()

      _cssCache = style.build()
    }
    return _cssCache
  }

  function clearCache() {
    style = new StyleSheet()
    _cssCache = undefined

    const preflightSafelist = toArray(preflightOptions?.safelist || []).flatMap(i => i.split(' '))

    include(classesPending, configSafelist)
    include(classesPending, safelist)
    include(classesPending, classesGenerated)

    include(tagsPending, tagsGenerated)
    include(tagsPending, preflightTags)
    include(tagsPending, preflightSafelist)
    include(tagsAvailable, htmlTags as any as string[])

    exclude(tagsAvailable, preflightTags)
    exclude(tagsAvailable, preflightSafelist)

    classesGenerated.clear()
    tagsGenerated.clear()
  }

  function init() {
    processor = initWindicss()
    clearCache()
  }

  return {
    options,
    init,
    extractFile,
    generateCSS,
    clearCache,
    transformCSS,
    isDetectTarget,
    scan,

    classesGenerated,
    classesPending,
    tagsGenerated,
    tagsPending,
    tagsAvailable,

    get configFilePath() {
      return configFilePath
    },
  }
}
