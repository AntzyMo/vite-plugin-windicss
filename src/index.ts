import { promises as fs, existsSync } from 'fs'
import { join, resolve } from 'path'
import type { Plugin, ResolvedConfig } from 'vite'
import fg from 'fast-glob'
import Windicss from 'windicss'
import { StyleSheet } from 'windicss/utils/style'
import { Config as WindiCssOptions } from 'windicss/types/interfaces'
import { MODULE_ID, MODULE_ID_VIRTUAL, preflightHTML } from './constants'
import { debug } from './debug'
import { Options } from './types'

function VitePluginWindicss(options: Options = {}): Plugin[] {
  const {
    windicssOptions = 'tailwind.config.js',
    searchExtensions = ['html', 'vue', 'pug'],
    searchDirs = ['src'],
    preflight = true,
  } = options

  let config: ResolvedConfig
  let windi: Windicss
  let windiConfigFile: string | undefined
  const extensionRegex = new RegExp(`\\.(?:${searchExtensions.join('|')})$`, 'i')

  const names = new Set<string>()
  let ignored = new Set<string>()

  let preflightStyle: StyleSheet | undefined
  const preflightOptions = Object.assign({
    html: preflightHTML,
    includeBase: true,
    includeGlobal: true,
    includePlugin: true,
  }, typeof preflight === 'boolean' ? {} : preflight)

  function createWindicss() {
    let options: WindiCssOptions = {}
    if (typeof windicssOptions === 'string') {
      const path = resolve(config.root, windicssOptions)
      if (!existsSync(path)) {
        console.warn(`[vite-plugin-windicss] config file "${windicssOptions}" not found, ignored`)
      }
      else {
        try {
          delete require.cache[require.resolve(path)]
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          options = require(path)
          windiConfigFile = path
        }
        catch (e) {
          console.error(`[vite-plugin-windicss] failed to load config "${windicssOptions}"`)
          console.error(`[vite-plugin-windicss] ${e.toString()}`)
          process.exit(1)
        }
      }
    }
    else {
      options = windicssOptions
    }

    debug.config(JSON.stringify(options, null, 2))
    return new Windicss(options)
  }

  let _searching: Promise<void> | null

  async function search() {
    if (!_searching) {
      _searching = (async() => {
        const files = await fg(
          searchDirs.map(i => join(i, `**/*.{${searchExtensions.join(',')}}`)),
          {
            onlyFiles: true,
            cwd: config.root,
            absolute: true,
          })

        debug.glob(files)

        await Promise.all(files.map(async(id) => {
          const content = await fs.readFile(id, 'utf-8')
          detectFile(content, id)
        }))
      })()
    }

    return _searching
  }

  function isDetectTarget(id: string) {
    return id.match(extensionRegex)
  }

  function detectFile(code: string, id: string) {
    if (!isDetectTarget(id))
      return

    debug.detect(id)
    Array.from(code.matchAll(/['"]([\w -/:]+)['"]/g))
      .flatMap(([, i]) => i.trim().split(' '))
      .forEach((i) => {
        if (ignored.has(i))
          return
        names.add(i)
      })
    debug.detect(names)
  }

  return [
    {
      name: 'vite-plugin-windicss:pre',
      enforce: 'pre',

      configResolved(_config) {
        config = _config
        windi = createWindicss()
      },

      resolveId(id): string | null {
        return id.startsWith(MODULE_ID) || id === MODULE_ID_VIRTUAL
          ? MODULE_ID_VIRTUAL
          : null
      },

      async load(id) {
        if (id === MODULE_ID_VIRTUAL) {
          await search()
          if (preflight && !preflightStyle) {
            preflightStyle = windi.preflight(
              preflightOptions.html,
              preflightOptions.includeBase,
              preflightOptions.includeGlobal,
              preflightOptions.includePlugin,
            )
          }

          const result = windi.interpret(Array.from(names).join(' '))
          result.ignored.forEach((i) => {
            names.delete(i)
            ignored.add(i)
          })
          debug.compile(`compiling ${names.size} classes`)
          let style = result.styleSheet
          if (preflightStyle)
            style = style.extend(preflightStyle, true)

          const css = style.build()
          return css
        }
      },
    },
    {
      name: 'vite-plugin-windicss:hmr',
      apply: 'serve',
      enforce: 'post',

      configureServer(server) {
        if (windiConfigFile)
          server.watcher.add(windiConfigFile)
      },

      async handleHotUpdate({ server, file, read, modules }) {
        if (windiConfigFile && file === windiConfigFile) {
          debug.hmr(`config file changed: ${file}`)
          windi = createWindicss()
          Array.from(ignored.values()).forEach(i => names.add(i))
          ignored = new Set()
          setTimeout(() => {
            console.log('[vite-plugin-windicss] configure file changed, reloading')
            server.ws.send({ type: 'full-reload' })
          }, 0)
          return [server.moduleGraph.getModuleById(MODULE_ID_VIRTUAL)!]
        }

        if (!isDetectTarget(file))
          return

        debug.hmr(`refreshed by ${file}`)

        detectFile(await read(), file)

        const module = server.moduleGraph.getModuleById(MODULE_ID_VIRTUAL)
        return [module!, ...modules]
      },
    },
  ]
}

export default VitePluginWindicss
