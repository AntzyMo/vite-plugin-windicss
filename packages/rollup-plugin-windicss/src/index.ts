import { Plugin } from 'rollup'
import _debug from 'debug'
import { UserOptions, WindiPluginUtils } from '@windicss/plugin-utils'

const NAME = 'rollup-plugin-windicss'
const MODULE_ID = 'windi.css'
const MODULE_ID_VIRTUAL = `/@windicss/${MODULE_ID}`

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
}

function WindiCssRollupPlugin(options: UserOptions = {}): Plugin[] {
  let utils: WindiPluginUtils

  const plugins: Plugin[] = []

  if (options.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      transform(code, id) {
        if (!utils.isScanTarget(id))
          return
        debug.group(id)
        return utils.transfromGroups(code)
      },
    })
  }

  plugins.push({
    name: `${NAME}:entry`,

    resolveId(id): string | null {
      return id.startsWith(MODULE_ID) || id === MODULE_ID_VIRTUAL
        ? MODULE_ID_VIRTUAL
        : null
    },

    async load(id) {
      if (id === MODULE_ID_VIRTUAL)
        return utils.generateCSS()
    },
  })

  if (options.transformCSS !== false) {
    plugins.push({
      name: `${NAME}:css`,
      transform(code, id) {
        if (!utils.isCssTransformTarget(id))
          return
        debug.css(id)
        return utils.transformCSS(code)
      },
    })
  }

  return plugins
}

export * from '@windicss/plugin-utils'
export default WindiCssRollupPlugin
