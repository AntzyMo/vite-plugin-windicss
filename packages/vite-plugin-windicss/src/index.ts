import { Plugin } from 'vite'
import _debug from 'debug'
import { UserOptions, createUtils, WindiPluginUtils, transfromGroups } from '@windicss/plugin-utils'

const NAME = 'vite-plugin-windicss'
const MODULE_ID = 'windi.css'
const MODULE_ID_VIRTUAL = `/@windicss/${MODULE_ID}`

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
}

function VitePluginWindicss(options: UserOptions = {}): Plugin[] {
  let utils: WindiPluginUtils

  const plugins: Plugin[] = []

  // Utilities grouping transform
  if (options.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      transform(code, id) {
        if (!utils.isScanTarget(id))
          return
        debug.group(id)
        return transfromGroups(code)
      },
    })
  }

  // CSS Entry via virtual module
  plugins.push({
    name: `${NAME}:pre`,
    enforce: 'pre',

    configResolved(_config) {
      utils = createUtils({
        ...options,
        _pluginName: NAME,
        _projectRoot: _config.root,
      })
      utils.init()
    },

    resolveId(id): string | null {
      return id.startsWith(MODULE_ID) || id === MODULE_ID_VIRTUAL
        ? MODULE_ID_VIRTUAL
        : null
    },

    async load(id) {
      if (id === MODULE_ID_VIRTUAL)
        return utils.generateCSS()
    },
  })

  // HMR
  plugins.push({
    name: `${NAME}:hmr`,
    apply: 'serve',
    enforce: 'post',

    configureServer(server) {
      if (utils.configFilePath)
        server.watcher.add(utils.configFilePath)
    },

    async handleHotUpdate({ server, file, read, modules }) {
      if (file === utils.configFilePath) {
        debug.hmr(`config file changed: ${file}`)
        utils.init()
        setTimeout(() => {
          console.log(`[${NAME}] configure file changed, reloading`)
          server.ws.send({ type: 'full-reload' })
        }, 0)
        return [server.moduleGraph.getModuleById(MODULE_ID_VIRTUAL)!]
      }

      if (!utils.isDetectTarget(file))
        return

      debug.hmr(`refreshed by ${file}`)

      utils.extractFile(await read())

      const module = server.moduleGraph.getModuleById(MODULE_ID_VIRTUAL)!
      server.moduleGraph.invalidateModule(module)

      if (file.endsWith('.html'))
        return undefined

      return [module!, ...modules]
    },
  })

  // CSS transform
  if (options.transformCSS !== false) {
    plugins.push({
      name: `${NAME}:css`,
      transform(code, id) {
        if (!utils.isCssTransformTarget(id))
          return
        debug.css(id)
        return utils.transformCSS(code)
      },
    })
  }

  return plugins
}

export * from '@windicss/plugin-utils'
export default VitePluginWindicss
