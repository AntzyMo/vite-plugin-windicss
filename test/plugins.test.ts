import { createBox } from '../src/box'

describe('plugins', () => {
  it('aspect-ratio', async() => {
    const box = createBox({
      windicssOptions: {
        plugins: [
          require('windicss/plugin/aspect-ratio'),
        ],
      },
      scan: false,
      preflight: false,
    })
    box.init()
    box.extractFile('"aspect-none aspect-w-16 aspect-h-9 aspect-16/9"')

    const css = await box.generateCSS()
    expect(box.classesGenerated).toMatchSnapshot('classes')
    expect(css).toMatchSnapshot('generated-css')
  })
})
