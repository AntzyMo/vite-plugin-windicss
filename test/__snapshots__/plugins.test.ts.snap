// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins aspect-ratio: classes 1`] = `
Set {
  "aspect-none",
  "aspect-w-16",
  "aspect-h-9",
  "aspect-16/9",
}
`;

exports[`plugins aspect-ratio: generated-css 1`] = `
".aspect-none {
  position: static;
  padding-bottom: 0;
}
.aspect-none > * {
  position: static;
  height: auto;
  width: auto;
  top: auto;
  right: auto;
  bottom: auto;
  left: auto;
}
.aspect-w-16 {
  --tw-aspect-w: 16;
  position: relative;
  padding-bottom: calc(var(--tw-aspect-h) / var(--tw-aspect-w) * 100%);
}
.aspect-w-16 > * {
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}
.aspect-h-9 {
  --tw-aspect-h: 9;
}
.aspect-16\\\\/9 {
  position: relative;
  padding-bottom: 177.777778%;
}
.aspect-16\\\\/9 > * {
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}"
`;
